pkg/api/login.go

```diff
@@ -0,0 +1,87 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/virtual-vgo/vvgo/pkg/discord"
+	"github.com/virtual-vgo/vvgo/pkg/login"
+	"github.com/virtual-vgo/vvgo/pkg/tracing"
+	"net/http"
+	"time"
+)
+
+const DiscordLoginExpires = 7 * 24 * 3600 * time.Second // 1 week
+
+// DiscordLoginHandler accepts an oauth token in the request body and uses the token to query for discord identity.
+// If the discord identity is a member of the vvgo discord server and has the vvgo-member role,
+// authentication is established and a login session cookie is sent in the response.
+// Otherwise, 401 unauthorized.
+type DiscordLoginHandler struct {
+	GuildID        discord.GuildID
+	RoleVVGOMember string
+	Discord        *discord.Client
+	Sessions       *login.Store
+}
+
+var ErrNotAMember = errors.New("not a member")
+
+func (x DiscordLoginHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx, span := tracing.StartSpan(r.Context(), "discord_oauth_handler")
+	defer span.Send()
+
+	handleError := func(err error) bool {
+		if err != nil {
+			tracing.AddError(ctx, err)
+			logger.WithError(err).Error("discord authentication failed")
+			unauthorized(w)
+			return false
+		}
+		return true
+	}
+
+	// read the oauth token from the request
+	var oauthToken discord.OAuthToken
+	if ok := handleError(json.NewDecoder(r.Body).Decode(&oauthToken)); !ok {
+		return
+	}
```

Comment:
Honestly, I'd rather just store the token in Redis, or just have this take a Discord User Snowflake from Redis. Either way, it'd be attached to the user's pre-session map.

All of this gets tricky once we start worrying about this plus Username/Pass authentication, though.